From a85e626da8ab27fa942bbd06bcf5f02878a066af Mon Sep 17 00:00:00 2001
From: Matthew Ng <ngmatthew@google.com>
Date: Tue, 7 Nov 2017 11:50:36 -0800
Subject: [PATCH 013/306] Fixes service disconnect crash when changing users

Changing users crashes system ui because service disconnected from
unprovisioned user that had no connection. Fix by always disconnecting
if connected before trying to connect to service. Requires follow up
change to have new provisioned user to connect to service.

Change-Id: I350ec45b90fa297c6ffe4932325c3b07653538f3
Fixes: 68929438
Test: manual
(cherry picked from commit 1fa3f7eebc0deb8170a1a57422af2e9cb8f9bafd)
---
 .../com/android/systemui/OverviewProxyService.java    | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/OverviewProxyService.java b/packages/SystemUI/src/com/android/systemui/OverviewProxyService.java
index 3878cd1cf92..4194701327f 100644
--- a/packages/SystemUI/src/com/android/systemui/OverviewProxyService.java
+++ b/packages/SystemUI/src/com/android/systemui/OverviewProxyService.java
@@ -78,7 +78,6 @@ public class OverviewProxyService {
 
             @Override
             public void onUserSwitched() {
-                disconnectFromLauncherService();
                 mConnectionBackoffAttempts = 0;
                 startConnectionToCurrentUser();
             }
@@ -100,8 +99,10 @@ public class OverviewProxyService {
     }
 
     public void startConnectionToCurrentUser() {
+        disconnectFromLauncherService();
+
         // If user has not setup yet or already connected, do not try to connect
-        if (!mDeviceProvisionedController.isCurrentUserSetup() || mOverviewProxy != null) {
+        if (!mDeviceProvisionedController.isCurrentUserSetup()) {
             return;
         }
         mHandler.removeCallbacks(mConnectionRunnable);
@@ -124,7 +125,9 @@ public class OverviewProxyService {
     }
 
     private void disconnectFromLauncherService() {
-        mContext.unbindService(mOverviewServiceConnection);
-        mOverviewProxy = null;
+        if (mOverviewProxy != null) {
+            mContext.unbindService(mOverviewServiceConnection);
+            mOverviewProxy = null;
+        }
     }
 }
-- 
2.17.1

